sinatraを使って、メモアプリを作成

sochi419 opened the pull request.

sochi419 committed:
first commit

sochi419 commented:
sinatraでメモアプリを作成しました!

確認よろしくお願いします🙏

sochi419 committed:
Changed the following points
・ xss processing

sochi419 commented:
xss処理の挙動で正しくエスケープ処理ができていないことが判明したので、修正しました。
また、変数名が好ましくないと感じたので、変数名を変えました。

よろしくお願いします🙏

sochi419 commented on memo.rb:
```diff
@@ -0,0 +1,101 @@
+    hash = JSON.load(file)
+    @hashes = hash['memos']
+  end
+end
```
rubocopにて、`JSON.load(f)`の書き方は好ましくないと警告されたのですが、修正必須でしょうか？（動作は正常です）
 
 現場では`JSON.parse`の書き方が主流なのでしょうか？

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,101 @@
+
+post '/create' do
+  # 新規メモの、ID番号を決める処理。
+  id_aggregation = []
+
+  @memos['memos'].each do |memo|
+    id_aggregation << memo['id']
+  end
```
```ruby
array = []
source_array.each do |item|
  array << item.xxx
end
```
というパターンは
```ruby
array = source_array.map { |item| item,xxx }
```
と `map` を使えばより簡単に書けます。
「配列の各要素に何かしらの処理をして別の配列にする」ときは `map` をうまくつかえるとよいかなと思います
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/collect.html

sochi419 replied:
仰るとおり、mapメソッドを使った方が簡潔に書けますね！
````
id_aggregation =  @memos['memos'].map { |memo| memo["id"] }
````
に変更しました〜

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,101 @@
+    id_aggregation << memo['id']
+  end
+
+  # 「既存メモの最大ID + 1」 に新規メモのIDを設定する。
+  new_memo_id = (id_aggregation.max.to_i + 1)
```
今回はこれでもOKですが、実際にはこれだと問題があります。

- 通常Webサーバーは複数台で運用されるため、同時にアクセスが発生する。同時アクセス時には新規発行idが重複してしまう可能性がある。
- あるidのメモを削除後に別のメモが同じidで作られる可能性がある。これはもしURLでブックマークなどしているユーザーがいると混乱の元になります。

ではどうするかというと、

- 同時アクセスを制御する何らかのしくみ(ロックなど)を使って、連番のidを発行する。一度発行されたidは二度と発行しないようにする(必ず前回発行したidより大きいidしか発行しない)
- uuidなどランダムなidを使う

などの対策があります。

yoshitsugu commented on views/show.erb:
```diff
@@ -0,0 +1,21 @@
+<p></p>
+
+<form method="post" action="/show/<%= h(@hash["id"]) %>" >
+<input type="hidden" name="_method" value="DELETE">
```
これだけだと多分DELETEリクエストはおくられなくて、何らかのjsのサポートとか必要になるかと思うんですが、どうでしょうか？

sochi419 replied:
`memo.rb`の69行目の`@hashes.delete(@result)`において、変数名を適切なものに変えようとした時に、誤ってこの1文だけ、変更前の変数名のままだったようです。（ただの見落としでした、申し訳ありません🙏）

こちらに関しては正しい変数名に再度修正をして、削除機能が機能していることを確認しました。

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,101 @@
+get '/show/:id' do
+  # 複数のメモの中で、表示したいメモ(idが一致するもの)を@hashに代入する。
+  keyword = params['id'].to_s
+  @hash = @hashes.find { |x| x['id'].match?(keyword) }
```
これmatchだとマズくないですか？
https://docs.ruby-lang.org/ja/latest/method/String/i/match=3f.html
`keyword` 側が勝手に正規表現になっちゃうので、たとえば `id: 1` と `id: 10` のメモがあると `keyword = 1` では両方マッチしちゃいますね。

sochi419 replied:
match?(条件) は、完全一致の条件だと思い込んでいました。
部分一致なんですね。勉強になります。

match?に該当する条件の部分を、
`memo['id'].to_i == params['id'].to_i`
としました。

````
@memo_infos.each do |memo| 
    @memo_info = (memo['id'].to_i == params['id'].to_i) ? memo : @memo_info
  end  
````

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,101 @@
+
+  File.open('memo.json') do |file|
+    hash = JSON.load(file)
+    @hashes = hash['memos']
```
`hashes` という名前は実体をあらわしていないので、あまりよくないですね。名前づけ工夫してみてください〜

sochi419 replied:
`hashes`の中身は`[{"id"=>"1", "title"=>"\"<メモ１>\"", "body"=>"メモ1を新たに作成しました。"}]`のように
id, タイトル, 本文が、1つの配列に格納されているハッシュ群なので、`memo_infos`という名前にしてみました。

````
before do
  @memos = File.open('memo.json') { |f| JSON.load(f) }

  File.open('memo.json') do |file|
    @memo_infos = (JSON.load(file))['memos']
  end
end
````

yoshitsugu commented on views/index.erb:
```diff
@@ -0,0 +1,15 @@
+    <% @hashes.each do |hash| %>
+        <tr>
+            <td><%= hash["id"] %></td>
+            <td><a href='/show/<%= hash["id"] %>'><%=  h(hash["title"]) %></a></td>
```
`memo,rb` でもエスケープしているのに、ここでもエスケープしているので、完全に入力と表示がかわってしまっていますね。view側だけにしたほうがいいかなと思います。

sochi419 replied:
view側だけに変更しました。

yoshitsugu reviewed (commented):
気になったところコメントしました！

path設計についてはRESTのプラクティスを思い出してみてください〜

sochi419 commented on memo.rb:
```diff
@@ -0,0 +1,101 @@
+  File.open('memo.json', 'w') do |file|
+    keyword = params['id'].to_s
+    @hash = @hashes.find { |x| x['id'].match?(keyword) }
+    @hashes.delete(@result)
```
変数名を変えた時に、一箇所だけ前の変数名のままにしていたため、
削除機能が機能していませんでした。（69行目の`@result`という変数名が、変更前のまま）

変数名を正しいものに変えました。

sochi419 committed:
Fixed the following points
--Fixed variable name
--Fixed escape processing
--Fixed delete method
--Fixed match? method
--Changed to process using map method

sochi419 commented:
以下の点を修正しました!

- 変数名の修正
- エスケープ処理の修正
- deleteメソッドの修正
- match?メソッドの修正
- mapメソッドを使った処理に変更

前回のPRで頂いたレビューに対しての返信で、変更点の説明をしています。

確認よろしくお願いします🙏

yoshitsugu commented:
@sochi419 
> 前回のPRで頂いたレビューに対しての返信で、変更点の説明をしています。

コメントが見えないので、pendingになっているかもです〜
参考: https://bootcamp.fjord.jp/pages/91#git-hub-pull-request

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,93 @@
+  erb :index
+end
+
+get '/show/:id' do
```
RESTのプラクティスを思い出してパスを変更してみてください〜

sochi419 replied:
`web技術の基本`を読み直して、自分のURI設計で良くないところを考えてみました。
URI設計で、動詞を使うのは好ましくないと記述があったので、
`show`を`detail`に変更しました。
`/show/:id` → `/detail/:id`

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,93 @@
+end
+
+before do
+  @memos = File.open('memo.json') { |f| JSON.load(f) }
```
rubocopの件、回答してなかったですね。すいません。
parseには簡単に変更できますので、やってみてください！
https://docs.ruby-lang.org/ja/latest/method/JSON/m/parse.html
parseはloadと違って、JSONの文字列自体を入力にすればいいです。
なので、ファイルオブジェクトではなく、ファイルの中身の文字列を渡してあげれば大丈夫です。

sochi419 replied:
parseを使った記法に変更しました。
現場では、loadは使わずparseを使用すべきなのでしょうか？

yoshitsugu replied:
@sochi419
`JSON.load` のリファレンスマニュアルを見てもらうと
https://docs.ruby-lang.org/ja/latest/method/JSON/m/load.html
sourceとして、 `JSON 形式の文字列を指定します。他には、to_str, to_io, read メソッドを持つオブジェクトも指定可能です。 ` と書いてありますね。これはいろいろなオブジェクトを `評価` して文字列とした上で、パースする、というメソッドです。ですので `to_str, to_io, read` としてメソッドを実装すれば、悪意のあるコードを実行されてしまう可能性があります。
たとえば `read` メソッドとしていろいろな場所のファイルを消してから "{}" を返す、というメソッドを持つクラスを設計することも可能です。
一方 `parse` は文字列をJSONとして解釈するだけ(解釈できないとエラー)なので、そういった余地はありません。
なので、 `load` でないとどうしても実装が難しい場合以外は `parse` を使っておくと、安全にコードがかけるかなと思います。もちろん今回のような場合は `load` するもの自体も最初から自分のコードで埋め込んでいるので、`parse` にしたから安全になるかというとそうでもないです。ただ、普段から `parse` を使うような癖をつけておくと「安全になりやすい」ですし、逆に `load` を使うときは変なコードが実行されてしまう設計になってしまわないか考える癖をつけておくのは大事です。

このあたり疑問を持つことは大事なので、この質問はとてもいい質問だと思います。次のステップとして、ご自身で調べて、「こういう解釈なんですが、あっていますか？」という形で質問できるとさらによいかと思います。是非やってみてください〜

sochi419 replied:
アドバイスありがとうございます！
参考記事を読んで、理解できました。
https://docs.ruby-lang.org/ja/latest/class/JSON.html

自分の`load`と`parse`の解釈
①`load`の方は、引数に`to_str, to_io, read`のような、オブジェクトを文字列に変換する処理を実装できる。そのため、悪意のあるコードが実行される可能性がある。

②`parse`は、`JSONの文字列`しか受け付けない（`to_str, to_io, read`のようなオブジェクトを文字列に変換する処理はエラー）ので、安全性の面で`load`より優れている。

> このあたり疑問を持つことは大事なので、この質問はとてもいい質問だと思います。次のステップとして、ご自身で調べて、「こういう解釈なんですが、あっていますか？」という形で質問できるとさらによいかと思います。是非やってみてください〜

そうですね。以後、質問の形式を「こういう解釈なんですが、あっていますか？」と言うふうに、自分の見解を述べてから質問できるよう、心がけたいと思います！

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,93 @@
+
+patch '/edit/:id' do
+  File.open('memo.json', 'w') do |file|
+    memo_inputs = [] # 配列memo_inputsに、メモの内容を入れていく。
+
+    @memos['memos'].each do |memo|
+      memo_inputs << if memo['id'] == params['id'].to_s # 編集中のメモは、ユーザーの入力内容を反映させた内容を、memo_inputsに代入。
+                       { 'id' => params['id'].to_s, 'title' => params['title'].to_s, 'body' => params['content'].to_s }
+                     else
+                       memo # 編集中ではないメモは、そのままmemo_inputsに代入。
+                     end
+    end
```
```ruby
array = []
source_array.each do |item|
  array << item.xxx
end
```
の形のものは

```ruby
array = original_array.map {|item| item.xxx }
```
と `map` を使うことで簡単に書けます。
配列を別の配列にするときはmapを思い出してもらえればと思います〜

sochi419 replied:
`.map`記法に変更しました。

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,93 @@
+
+delete '/show/:id' do
+  File.open('memo.json', 'w') do |file|
+    @memo_infos.each do |memo|
+      @memo_info = memo['id'].to_i == params['id'].to_i ? memo : @memo_info
+    end
+    @memo_infos.delete(@memo_info)
```
`delete_if` とかもあります〜
https://docs.ruby-lang.org/ja/latest/method/Array/i/delete_if.html

sochi419 replied:
`delete_if`を使った構文に変更しました。

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,93 @@
+before do
+  @memos = File.open('memo.json') { |f| JSON.load(f) }
+
+  File.open('memo.json') do |file|
+    @memo_infos = (JSON.load(file))['memos']
+  end
```
`memos` と `memo_infos` の違いがわかりにくいので、名前をもうちょっと考えてみてください〜。
そもそも `@memos` は `@memos['memos']` の形でしか参照されていないので、現状の `@memo_infos` だけでもよさそうな気がしますね :eyes:

sochi419 replied:
仰る通り、`memos`と`memo_infos`の中身は似通っているので、どちらか一つで良さそうですね〜

`@memo_infos`のみにしました!

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,93 @@
+end
+
+get '/show/:id' do
+  @memo_infos.each do |memo|
+    @memo_info = memo['id'].to_i == params['id'].to_i ? memo : @memo_info
+  end
```
`detect` とかもうまくつかってみてくださいね。
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/detect.html

sochi419 replied:
findメソッドを使った記法に修正しました。

yoshitsugu reviewed (commented):
コメントしました！

sochi419 reviewed (commented):
`pending`で他者にコメントが見えない状態でした。
`submit review`をしたので、これで他者にも見えると思います。

sochi419 commented:
期間が空いてしまい、申し訳ありません。
指摘点を修正しました！

確認よろしくお願いします🙏

sochi419 committed:
Fixed the following points
--Change path name
--Change of json file reading process
--Changed to process using map method
--Changed to process using delete_if
--Change variable name
--Changed to process using find method

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,83 @@
+  erb :index
+end
+
+get '/detail/:id' do
```
ちょっとRESTなPath設計について復習してみましょう。
基本 `/(リソース名の複数系)` もしくは `/(リソース名の複数系)/(リソースID)` になっていましたよね
「ユーザーを表示する」Pathは `GET /users/:user_id` ですし、「リスト一覧を表示する」Pathは `GET /users/:user_id/lists` でした。「メモを表示する」Pathはどうなるでしょうか？

他のPathについても再度RESTのプラクティスを思い返して、リソースの追加、リソースの更新はどうなっていたか、考えてみてください！

sochi419 replied:
かなり前にrestの考え方について学習したので、URI設計の基本を見落としていました..
全体のURI設計を見直し、修正しました！

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,83 @@
+  File.open('memo.json', 'w') do |file|
+    memo_inputs = @memo_infos.map do |memo|
+      if memo['id'] == params['id'].to_s # 編集中のメモは、ユーザーの入力内容を反映させた内容を、memo_inputsに代入。
+        { 'id' => params['id'].to_s, 'title' => params['title'].to_s, 'body' => params['content'].to_s }
```
idは更新しちゃだめですよね。元のidをそのまま使いましょう。
`if` があるので、これによって問題はおこらないですが、変えなくていいものは変えないようにするとセキュリティ的にもよいかと思います。

sochi419 replied:
````
if memo['id'] == params['id'].to_s 
  { 'id' => memo['id'], 'title' => params['title'].to_s, 'body' => params['content'].to_s }
````
map処理のなかで、idは更新しないような処理にしました~

yoshitsugu reviewed (commented):
コメントしました！

sochi419 committed:
I changed the following points.
・Changed the path design.
・Changed to a process that does not update the id when editing a memo.

sochi419 commented:
修正しました！
確認よろしくお願いします🙏

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,87 @@
+  erb :edit
+end
+
+patch '/memos/:id/edit' do
```
`patch` Pathは通常 `/memos/:id` で十分かなと思います。

sochi419 replied:
patchのPathを、` /memos/:id` でに変更しました。

yoshitsugu commented on memo.rb:
```diff
@@ -0,0 +1,87 @@
+  if @memo_infos == []
+    new_memo_id = 1
+  else
+    id_aggregation = @memo_infos.map { |memo| memo['id'].to_i }
+    new_memo_id = (id_aggregation.max + 1) # 「既存メモの最大ID + 1」 に新規メモのIDを設定する。
```
`max_by` というのもあります。
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/max_by.html

sochi419 replied:
なるほど、`max_by`というものもあるんですね〜

`max_by`について調べたのですが、`max`ではなくわざわざ`max_by`を使う利点がよく分かりませんでした。

````
a = [1, 2, 3]
a.max                     # 3
a.max_by { |x| x }        # 3
````
のように、`max_by`はブロック変数{}を指定しなければいけない分、コードが冗長になる気がしてしまいました。

どんな場合に`max_by`を使うべきなのでしょうか？
（個人的には全部`max`でいいんじゃないかと、思ってしまいました。）

sochi419 replied:
`max_by`を使った処理に変更してみました。

yoshitsugu replied:
@sochi419 `max_by` を使うと、  `map` でやっている変換処理も含めて処理できるので、たとえば「2次元座標のX座標の最大値をもつ点とりたい」というときに

```ruby
point1 = [1, 2]
point2 = [3, 5]
point3 = [2, 10]

[point1, point2, point3].max_by { |point| point[0] }
#=> [3, 5]
```

みたいにできます。今回みたいに、最大値自体がほしい場合はどちらでもあまりかわらないかもしれないですね。
一応以下のようにできるとは思います。

```ruby
# 結局2回 ['id'].to_i しないといけないのであまり効率よくないかも。。。
new_memo_id = @memo_infos.max_by { |memo| memo['id'].to_i }['id'].to_i + 1
```

sochi419 replied:
説明ありがとうございます！

違いが良く理解できました！
違いを見比べてみました。
- `max`を使った場合
`new_memo_id = @memo_infos.map { |memo| memo['id'].to_i }.max + 1`

- `max_by`を使った場合
`new_memo_id = @memo_infos.max_by { |memo| memo['id'].to_i }['id'].to_i + 1`

maxを使った場合は、**mapで変換処理をしてから**、maxしている。
一方max_byを使った場合は、**mapの変換処理を含んでいる**ので、`map`と`max`を合わせた処理を、max_byのみで事足りる。ということなんですね!

今回は`max_by`についての知見を得れたので、
`max_by`を使った処理で、提出したいと思います!

yoshitsugu reviewed (commented):
細かいところコメントしました！

sochi419 committed:
Fixed the following points
・Change path name
・Change to processing using max_by

sochi419 commented:
指摘点を修正しました！
確認よろしくお願いします🙏

sochi419 committed:
Change to max_by processing

sochi419 commented:
疑問点について、解説いただきありがとうございました！
`max_by`を使った処理の利点について、理解することができました🙌

yoshitsugu reviewed (approved):
OKです！
